contracts/terminal_value_oracle/main.cairo

```diff
@@ -0,0 +1,264 @@
+        new_value,
+        updater_address
+    );
+
+    last_update.write(
+        request,
+        new_update
+    );
+    
+    return();
+}
+
+// Function for registering new Request
+@external 
+func register_request{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    maturity: felt,
+    requested_address: felt,
+    reward: Uint256
+) {
+    alloc_locals;
+
+    // Assert that requested maturity hasn't already expired
+    let (current_block_time) = get_block_timestamp();
+    with_attr error_message("Can't setup Request with expired maturity") {
+        assert_le(maturity, current_block_time - 1);
+    }
+
+    // Create new Request
+    let request = Request (
+        maturity,
+        requested_address,
+        reward
+    );
+
+    // Get usable index for new request and write it there
+    let (usable_idx) = get_requests_usable_index(0);
+
+    requests.write(
+        usable_idx,
+        request
+    );
+
+    // Move reward from the caller to contract
+    let (caller_address) = get_caller_address();
+    let (own_address) = get_contract_address();
+    IERC20.transferFrom(
+        contract_address=ETH_ADDRESS,
+        sender = caller_address,
+        recipient = own_address,
+        amount = reward
+    );
+
+    // Update contract_balance storage_var
+    let (current_balance) = contract_balance.read();
+    let (new_balance, _) = uint256_add(current_balance, reward);
+    contract_balance.write(new_balance);
+        
+    return ();
+}
+
+// Function for cashing out, used by the last updater after 
+// the Request has expired
+@external
+func cashout_last_update{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(idx: felt) {
+    alloc_locals;
+
+    // Read Request at provided index
+    let (request) = requests.read(idx);
+
+    // Assert that Request has already expired
+    let (current_block_time) = get_block_timestamp();
+    with_attr error_message("Request isn't expired yet") {
+        assert_le(request.maturity, current_block_time);
+    }
+
+    // Assert that caller is the last updater
+    let (latest_update) = last_update.read(request);
+    let (caller_address) = get_caller_address();
+    with_attr error_message("Caller isn't the last updater"){
+        assert caller_address = latest_update.updater_address;
+    }
+
+    // Pay the reward
+    // TODO: Deduct the transaction fee from reward
+    let (own_address) = get_contract_address();
+    IERC20.transfer(
+        contract_address = ETH_ADDRESS,
+        recipient = latest_update.updater_address,
+        amount = request.reward
+    );
+
+    // Update contract_balance var 
+    let (current_balance) = contract_balance.read();
+    let (new_balance) = uint256_sub(current_balance, request.reward);
+    contract_balance.write(new_balance);
+    
+    // Delete request, since it has been paid
+    // TODO: Make sure you can still read the last update even if the request is deleted
+    remove_request(idx);
```

Comment:
if the request gets removed, we can't interate over the values since for "latest_value" you need to know the "request" -> I would add additional storage_var "cashed_out_requests" or "matured_requests" based on how we design this one